### PR TITLE
Fix security issue: avoid directory interleaving vulnerabilities

### PR DESCRIPTION
**Pull request by jialez0**
Add legality checking of resource path to avoid reading files in parent directory.

cc @Xynnn007

### CODE REVIEW

**Review by jiazhang0 — APPROVED** [review]

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,37 +1,21 @@
-FROM rust:slim as builder
+FROM rust:latest as builder
 
 ENV DEBIAN_FRONTEND noninteractive
 
-RUN apt-get update && \
```
Comment (mythi) [inline docker/Dockerfile:5]: These `Dockerfile` changes are not related to the error or are they? They seem to revert my recent changes but it's not clear what's the error?

Comment (jialez0) [reply docker/Dockerfile:5]: Yes, I am studying the real reasons for CI errors, and I will remove the modifications here.

File: src/api/src/resource/mod.rs [hunk 2]
```diff
@@ -35,6 +35,18 @@ pub struct ResourceDesc {
     pub resource_tag: String,
 }
 
+impl ResourceDesc {
+    pub fn check_legality(&self) -> Result<()> {
```
Comment (sameo) [inline src/api/src/resource/mod.rs:39]:
```suggestion
    pub fn is_valid(&self) -> Bool {
```

A terminology nit: legality refers to the actual law (as in judges and lawyers :) ). And I think returning a boolean seems more natural?

File: src/api/src/resource/mod.rs [hunk 3]
```diff
@@ -35,6 +35,18 @@ pub struct ResourceDesc {
...
+        if &self.repository_name == "." || &self.repository_name == ".." {
+            bail!(anyhow!("Illegal resource repository name"));
```
Comment (sameo) [inline src/api/src/resource/mod.rs:41]: Similar to the previous comment:

```suggestion
            bail!(anyhow!("Invalid resource repository name"));
```

File: src/api/src/resource/mod.rs [hunk 4]
```diff
@@ -35,6 +35,18 @@ pub struct ResourceDesc {
...
+        }
+        if &self.resource_type == "." || &self.resource_type == ".." {
```
Comment (fitzthum) [inline src/api/src/resource/mod.rs:43]: I assume we have already parsed out any slashes by this point?

Comment (jialez0) [reply src/api/src/resource/mod.rs]: es, the Actix library will help us complete the parsing of slashes when parsing the path information of URL.
